synapse/storage/databases/main/event_push_actions.py

```diff
@@ -1096,13 +1212,18 @@ def _handle_new_receipts_for_notifs_txn(self, txn: LoggingTransaction) -> bool:
         )
         rows = txn.fetchall()
 
-        # For each new read receipt we delete push actions from before it and
-        # recalculate the summary.
-        for _, room_id, user_id, stream_ordering in rows:
+        # First handle all the rows without a thread ID (i.e. ones that apply to
+        # the entire room).
+        for _, room_id, user_id, thread_id, stream_ordering in rows:
             # Only handle our own read receipts.
             if not self.hs.is_mine_id(user_id):
                 continue
 
+            if thread_id is not None:
+                continue
```

Comment:
Maybe it would be clearer to partition rows up front into `unthreaded_receipts` and `threaded_receipts`?

Theoretically you also don't care about any threaded receipts with an earlier stream ordering than the unthreaded one in the room.

Reply:
I ended up combining this into a single loop.